src/query/sortResults.ts

```diff
@@ -0,0 +1,89 @@
+import debug from 'debug'
+import get from 'lodash.get'
+import { Entity, InternalEntity } from 'src/glossary'
+import { OrderBy, SortDirection } from './queryTypes'
+
+const log = debug('sortResults')
+
+type FlatSortCriteria = [string[], SortDirection]
+
+function warnOnIneffectiveSortingKeys(sortCriteria: Record<string, any>): void {
+  const [mainCriteria, ...siblings] = Object.keys(sortCriteria)
+
+  if (siblings.length > 0) {
+    console.warn(
+      'Sorting by "%s" has no effect: already sorted by "%s".',
+      siblings.join(','),
+      mainCriteria,
+    )
+  }
+}
+
+function flattenSortCriteria<EntityType extends Entity<any, any>>(
+  orderBy: OrderBy<EntityType>[],
+  propertyPath: string[] = [],
+): FlatSortCriteria[] {
+  log('flattenSortCriteria:', orderBy, propertyPath)
+
+  return orderBy.reduce<FlatSortCriteria[]>((criteria, properties) => {
+    warnOnIneffectiveSortingKeys(properties)
+
+    // Multiple properties in a single criteria object are forbidden.
+    // Use the list of criteria objects for multi-criteria sort.
+    const property = Object.keys(properties)[0] as keyof OrderBy<EntityType>
+    const sortDirection = properties[property]!
```

Comment:
hm, I would avoid `!` here, it can be really `undefined` in runtime, e.g. if a user saves in the middle of editing:`[{}, {second: 'asc'}]`